prototypes/alpaka_binding/alpaka/README.md

```diff
@@ -0,0 +1,48 @@
+# About
+
+The prototype uses pybind11 to create a Python 3 binding from Alpaka C++ code. The goal of the implementation is to share memory between the Python and Alpaka C++ code and to implement an algorithm in CUDA.
+
+This is the last step in providing a Python binding of an algorithm written with the alpaka library.
+
+# Requirements
+
+* pybind11
+* numpy
+* cmake > 3.14
+* cuda > 9.0
+
+Available via pip or conda.
+
+```bash
+  # use CUDA of the host system
+  conda install -c conda-forge cmake=3.16 pybind11
+  conda install -c anaconda numpy boost
+```
+
+Install Alpaka
+
+```bash
+#cd ptychography/prototypes/alpaka_binding/
+git https://github.com/ComputationalRadiationPhysics/alpaka.git
+mkdir alpaka/build && cd alpaka/build
+cmake .. -DCMAKE_INSTALL_PREFIX=${CONDA_PREFIX} -Dalpaka_BUILD_EXAMPLES=OFF -DBUILD_TESTING=OFF
+make -j install
+```
+
+# Install
+
+```bash
+  mkdir build
+  cd build
+  # maybe you need to set: -DPYTHON_EXECUTABLE:FILEPATH=
+  cmake ..
+  make
```

Comment:
I got an error that seems to be C++-related:

```
(libertem) C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build>msbuild ALL_BUILD.vcxproj
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 09/04/2020 14:02:22.
Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.vcxproj" on no
de 1 (default targets).
Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.vcxproj" (1) i
s building "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ZERO_CHECK.vcxproj" (
2) on node 1 (default targets).
PrepareForBuild:
  Creating directory "x64\Debug\ZERO_CHECK\".
  Creating directory "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".
InitializeBuildStatus:
  Creating "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
CustomBuild:
  Checking Build System
FinalizeBuildStatus:
  Deleting file "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".
  Touching "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".
Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ZERO_CHECK
.vcxproj" (default targets).

Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.vcxproj" (1) i
s building "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.vcxproj
" (3) on node 1 (default targets).
PrepareForBuild:
  Creating directory "alpakaBinding.dir\Debug\".
  Creating directory "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\Debug\".
  Creating directory "alpakaBinding.dir\Debug\alpakaBinding.tlog\".
InitializeBuildStatus:
  Creating "alpakaBinding.dir\Debug\alpakaBinding.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
ComputeCustomBuildOutput:
  Creating directory "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\CMakeFiles\
  alpakaBinding.dir\Debug\".
CustomBuild:
  Building NVCC (Device) object CMakeFiles/alpakaBinding.dir/Debug/alpakaBinding_generated_binding.cpp.obj
  binding.cpp
c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1035): warning #186-D: pointless comparison of unsigned int
eger with zero [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.vcx
proj]
            detected during:
              instantiation of "__nv_bool pybind11::detail::type_caster<T, std::enable_if_t<<expression>, void>>::load(
  pybind11::handle, __nv_bool) [with T=pybind11::detail::intrinsic_t<size_t>]"
  (1973): here
              instantiation of "__nv_bool pybind11::detail::argument_loader<Args...>::load_impl_sequence(pybind11::deta
  il::function_call &, std::index_sequence<Is...>) [with Args=<pybind11::detail::value_and_holder &, size_t>, Is=<0ULL,
   1ULL>]"
  (1949): here
              instantiation of "__nv_bool pybind11::detail::argument_loader<Args...>::load_args(pybind11::detail::funct
  ion_call &) [with Args=<pybind11::detail::value_and_holder &, size_t>]"
  C:\Users\weber\.conda\envs\libertem\include\pybind11/pybind11.h(141): here
              instantiation of "void pybind11::cpp_function::initialize(Func &&, Return (*)(Args...), const Extra &...)
   [with Func=lambda [](pybind11::detail::value_and_holder &, size_t)->void, Return=void, Args=<pybind11::detail::value
  _and_holder &, size_t>, Extra=<pybind11::name, pybind11::is_method, pybind11::sibling, pybind11::detail::is_new_style
  _constructor>]"
  C:\Users\weber\.conda\envs\libertem\include\pybind11/pybind11.h(72): here
              instantiation of "pybind11::cpp_function::cpp_function(Func &&, const Extra &...) [with Func=lambda [](py
  bind11::detail::value_and_holder &, size_t)->void, Extra=<pybind11::name, pybind11::is_method, pybind11::sibling, pyb
  ind11::detail::is_new_style_constructor>, <unnamed>=void]"
  C:\Users\weber\.conda\envs\libertem\include\pybind11/pybind11.h(1119): here
              instantiation of "pybind11::class_<type_, options...> &pybind11::class_<type_, options...>::def(const cha
  r *, Func &&, const Extra &...) [with type_=Algo<float, int>, options=<>, Func=lambda [](pybind11::detail::value_and_
  holder &, size_t)->void, Extra=<pybind11::detail::is_new_style_constructor>]"
  c:\users\weber\.conda\envs\libertem\include\pybind11\detail/init.h(175): here
              instantiation of "void pybind11::detail::initimpl::constructor<Args...>::execute(Class &, const Extra &..
  .) [with Args=<size_t>, Class=pybind11::class_<Algo<float, int>>, Extra=<>, <unnamed>=0]"
  C:\Users\weber\.conda\envs\libertem\include\pybind11/pybind11.h(1148): here
              instantiation of "pybind11::class_<type_, options...> &pybind11::class_<type_, options...>::def(const pyb
  ind11::detail::initimpl::constructor<Args...> &, const Extra &...) [with type_=Algo<float, int>, options=<>, Args=<si
  ze_t>, Extra=<>]"
  C:/Users/weber/Documents/LiberTEM/ptychography/prototypes/alpaka_binding/alpaka/binding.cpp(12): here
              instantiation of "void decare_algo_binding<ALGO_CLASS_T>(pybind11::module &, const std::string &) [with A
  LGO_CLASS_T=Algo<float, int>]"
  C:/Users/weber/Documents/LiberTEM/ptychography/prototypes/alpaka_binding/alpaka/binding.cpp(23): here

c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1495): error #852: expression must be a pointer to a comple
te object type [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.vcx
proj]

  1 error detected in the compilation of "C:/Users/weber/AppData/Local/Temp/tmpxft_00003cb0_00000000-8_binding.cpp4.ii"
  .
  binding.cpp
  CMake Error at alpakaBinding_generated_binding.cpp.obj.Debug.cmake:280 (message):
    Error generating file
    C:/Users/weber/Documents/LiberTEM/ptychography/prototypes/alpaka_binding/alpaka/build/CMakeFiles/alpakaBinding.dir/
  /Debug/alpakaBinding_generated_binding.cpp.obj


Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBind
ing.vcxproj" (default targets) -- FAILED.

Done Building Project "C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.
vcxproj" (default targets) -- FAILED.


Build FAILED.

"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.vcxproj" (default targ
et) (1) ->
"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.vcxproj" (default
target) (3) ->
(CustomBuild target) ->
  c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1035): warning #186-D: pointless comparison of unsigned i
nteger with zero [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.v
cxproj]


"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\ALL_BUILD.vcxproj" (default targ
et) (1) ->
"C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.vcxproj" (default
target) (3) ->
(CustomBuild target) ->
  c:\users\weber\.conda\envs\libertem\include\pybind11\cast.h(1495): error #852: expression must be a pointer to a comp
lete object type [C:\Users\weber\Documents\LiberTEM\ptychography\prototypes\alpaka_binding\alpaka\build\alpakaBinding.v
cxproj]

    1 Warning(s)
    1 Error(s)

Time Elapsed 00:00:37.31
```